File: source/_posts/translate_update_slugs.js
// translate_update_slugs.js
// Usage:
// 1. Install dependencies: npm install puppeteer
// 2. (Optional) Set BROWSER_PATH env var to full path of Chrome or Edge executable
//    e.g. setx BROWSER_PATH "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe"
// 3. Run: node translate_update_slugs.js

const fs = require('fs').promises;
const puppeteer = require('puppeteer');
const path = require('path');

const JSON_PATH = 'R:\\tzengshinfu.github.io\\source\\_posts\\rename_mapping.json';
const BROWSER_PATH = process.env.BROWSER_PATH || null; // set to Chrome/Edge exe if you want
const HEADLESS = false; // show browser so you can observe

function sleep(ms) { return new Promise(resolve => setTimeout(resolve, ms)); }

(async () => {
    try {
        const raw = await fs.readFile(JSON_PATH, 'utf8');
        const data = JSON.parse(raw);

        const launchOpts = { headless: HEADLESS, args: ['--start-maximized'] };
        if (BROWSER_PATH) launchOpts.executablePath = BROWSER_PATH;

        const browser = await puppeteer.launch(launchOpts);
        const page = await browser.newPage();
        await page.setViewport({ width: 1200, height: 900 });

        for (let i = 0; i < data.length; i++) {
            const obj = data[i];
            const title = obj.title || '';
            if (!title) {
                console.log(`[${i + 1}/${data.length}] skipping empty title for ${obj.old}`);
                continue;
            }

            const url = 'https://translate.google.com/?sl=auto&tl=en&text=' + encodeURIComponent(title) + '&op=translate';
            console.log(`[${i + 1}/${data.length}] Navigating to translate: ${title}`);
            await page.goto(url, { waitUntil: 'networkidle2' });

            try {
                await page.waitForSelector('span[jsname="W297wb"]', { timeout: 15000 });
                const translated = await page.$eval('span[jsname="W297wb"]', el => el.innerText.trim());
                // sanitize into slug-like string: remove punctuation, replace spaces with dashes
                let slug = translated.toLowerCase().replace(/[^a-z0-9\s-]/g, '').trim().replace(/\s+/g, '-');
                if (!slug) slug = translated.toLowerCase().trim();
                obj.slug = slug;
                console.log(`  -> translated: "${translated}" => slug: "${slug}"`);
            } catch (err) {
                console.warn(`  ! failed to extract translation for item ${i + 1}: ${err.message}`);
                // fallback: keep existing slug
            }

            // random delay 1~5 seconds
            const delay = Math.floor(Math.random() * 5000) + 1000;
            console.log(`  waiting ${delay}ms before next`);
            await sleep(delay);
        }

        // write back
        await fs.writeFile(JSON_PATH, JSON.stringify(data, null, 2), 'utf8');
        console.log('All done. File saved to', JSON_PATH);

        await browser.close();
    } catch (err) {
        console.error('Error:', err);
        process.exit(1);
    }
})();

File: source/_posts/translate_update_slugs.py
"""
translate_update_slugs.py

Reads rename_mapping.json, opens Chrome, visits Google Translate for each title,
extracts the translated English text, converts it to a slug-like string,
writes that into the object's "slug" property, and saves the JSON file.

Requirements:
  pip install selenium webdriver-manager

Usage:
  python translate_update_slugs.py

Notes:
  - The script uses webdriver-manager to auto-download the matching chromedriver.
  - If you want to use an existing Chrome installation, ensure Chrome is up-to-date.
"""

import json
import time
import random
import urllib.parse
from pathlib import Path

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, WebDriverException
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options

JSON_PATH = Path(r"R:\\tzengshinfu.github.io\\source\\_posts\\rename_mapping.json")


def slugify_from_text(text):
    if not text:
        return ""
    s = text.strip().lower()
    # remove characters except letters, numbers, spaces and hyphen
    import re

    s = re.sub(r"[^a-z0-9\s-]", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    # keep up to first 5 words
    parts = s.split()
    parts = parts[:100]
    return "-".join(parts)


def main():
    if not JSON_PATH.exists():
        print(f"JSON file not found: {JSON_PATH}")
        return

    data = json.loads(JSON_PATH.read_text(encoding="utf-8"))
    if not isinstance(data, list):
        print("Unexpected JSON structure: expected a list of objects.")
        return

    chrome_options = Options()
    # show browser so user can observe; change to True for headless
    chrome_options.headless = False
    chrome_options.add_argument("--start-maximized")

    try:
        service = Service(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=chrome_options)
    except WebDriverException as e:
        print("Failed to start Chrome webdriver:", e)
        return

    wait = WebDriverWait(driver, 20)

    try:
        for idx, obj in enumerate(data, start=1):
            slug = obj.get("slug", "")
            if slug:
                continue

            title = obj.get("title", "")
            if not title:
                print(f"[{idx}/{len(data)}] empty title, skipping (old={obj.get('old')})")
                continue

            url = "https://translate.google.com/?sl=auto&tl=en&text=" + urllib.parse.quote(title) + "&op=translate"
            print(f"[{idx}/{len(data)}] Navigating to translate: {title}")
            try:
                driver.get(url)
            except WebDriverException as e:
                print("  navigation failed:", e)
                continue

            translated = None
            try:
                wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, 'span[jsname="W297wb"]')))
                elem = driver.find_element(By.CSS_SELECTOR, 'span[jsname="W297wb"]')
                translated = elem.text.strip()
            except TimeoutException:
                print("  timeout waiting for translation element")
                # try alternative selector that sometimes holds translation
                try:
                    alt = driver.find_element(By.CSS_SELECTOR, "div.result-shield-container span")
                    translated = alt.text.strip()
                except Exception:
                    translated = None

            if translated:
                slug = slugify_from_text(translated)
                obj["slug"] = slug
                print(f'  -> translated: "{translated}" => slug: "{slug}"')
            else:
                print("  -> no translation extracted, leaving slug unchanged")

            delay = random.uniform(5, 10)
            print(f"  waiting {delay:.2f}s before next")
            time.sleep(delay)
    finally:
        try:
            driver.quit()
        except Exception:
            pass

    # write back
    JSON_PATH.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
    print("Saved updated JSON to", JSON_PATH)


if __name__ == "__main__":
    main()
